### ISSUE TITLE
Pre-commit hook proposal: filesize

### ISSUE THREAD

**Issue by lencioni** [brigade/overcommit#348]
It would be nice to have a hook that warned about large files being committed. 1MiB might be a good default, but it could probably even be set to something smaller.

**Comment by lencioni** [comment]
It would be nice to have a way to configure the hook to warn at one level and error at another level. For instance, if you wanted to warn at 512KiB but error at 1MiB, it would be nice if this hook could support that.

**Comment by jawshooah** [comment]
What units should we use for config values? Should we be able to parse `512KiB`, `512 KiB`, `1MB`, `1MiB`, etc? Or should it just be in bytes?

**Comment by lencioni** [comment]
I think we should stick with something simple, so I'd say that the unit should not be in the value (we can mention it in the config name for clarity). I'd expect most people to set these to something in the KiB->MiB range, so perhaps the value should be in KiB? What do you think?

**Comment by sds** [comment]
My perspective is probably skewed because I deal so much with configuring systems where memory/file size constraints need to be specified explicitly, but I _really_ prefer having units in any sort of measure of size. It's never intuitive what the default units are, and different tools use different defaults.

I would strongly prefer we support units (perhaps even enforcing that units are included). Whether we want to support the difference between `KB` and `KiB` I'm more flexible on (but personally: long live the [IEC](https://en.wikipedia.org/wiki/International_Electrotechnical_Commission)!)

I could see this being reused elsewhere, so it would make sense to go in the `Overcommit::Utils` module.

**Comment by lencioni** [comment]
@sds I think units are important but I was advocating them being used in the configuration name instead of the value. I don't really feel strongly either way.

**Comment by davidstosik** [comment]
If anyone get here and is interested in a solution, this is what we're gonna try to use where I work:
```rb
module Overcommit::Hook::PreCommit
  class FileSize < Base
    DEFAULT_SIZE_LIMIT_BYTES = 1_000_000 # 1MB

    def run
      return :pass if oversized_files.empty?

      oversized_files.map do |file|
        error_message_for(file)
      end
    end

    private

      def oversized_files
        @_oversized_files ||= build_oversized_file_list
      end

      def build_oversized_file_list
        applicable_files.select do |file|
          File.size(file) > size_limit
        end
      end

      def size_limit_bytes
        config.fetch("size_limit_bytes", DEFAULT_SIZE_LIMIT_BYTES)
      end

      def error_message_for(file)
        Overcommit::Hook::Message.new(
          :error,
          file,
          nil,
          error_text_for(file)
        )
      end

      def error_text_for(file)
        "#{relative_path_for(file)} is over the size limit."
      end

      def relative_path_for(file)
        Pathname.new(file).relative_path_from(repo_root_path)
      end

      def repo_root_path
        @_repo_root_path ||= Pathname.new(Overcommit::Utils.repo_root)
      end
  end
```

Obviously, this almost completely ignored the size unit conversation and went with Bytes.

This is what the config looks like:
```yml
PreCommit:
  FileSize:
    enabled: true
    size_limit_bytes: 1000000
```

And the resulting error:

![filesize](https://user-images.githubusercontent.com/816901/56025663-8531a900-5d4d-11e9-83ba-dd1e18184861.png)

**Comment by sds** [comment]
Awesome work, @davidstosik. 

My only suggestion is including the actual file size and the limit in the error message (so I know how "over" I am).

```
config/big_file.json is over the file size limit of 100000 bytes (is 200000 bytes)
```

Will merge if you include tests!

**Comment by davidstosik** [comment]
@sds Thanks for the comment! In case you missed it: https://github.com/brigade/overcommit/pull/636

**Comment by sds** [comment]
Implemented in #636.
